### PR TITLE
Update byte-pair-encoding.md

### PR DESCRIPTION
**Pull request by RHuidrom**
draft-1 for byte-pair-encoding.md

### CODE REVIEW

File: approaches/byte-pair-encoding.md [hunk 1]
```diff
@@ -5,3 +5,5 @@ layout: coming_soon
 title: Byte-pair encoding
 description: A data compression algorithm for machine translation
 ---
+
+Tokenisation is fundamental to understanding human language. Algorithms for tokenisation can be based on word, subword or character. In this article, we discuss about subword-based tokenisation followed by Byte-pair encoding (BPE) algorithm.
```
Comment (cefoo) [inline approaches/byte-pair-encoding.md:9]: Thanks for starting this article! When writing the rest of it, please remember that you should aim for [minimal content](https://machinetranslate.org/style#minimal-content).

Comment (bittlingmayer) [inline approaches/byte-pair-encoding.md:9]: - The first sentence should describe and define the topic - byte-pair encoding for machine translation - so for example "**Byte-pair encoding** (**BPE**) is a compression algorithm that is used for [tokenisation](/tokenisation) for neural machine translation."
- Don't define tokenisation here, just create the tokenisation article, then link to it
- Write like Wikipedia not an academic paper, so no "We discuss..."

File: approaches/byte-pair-encoding.md [hunk 2]
```diff
@@ -5,3 +5,5 @@ layout: coming_soon
 title: Byte-pair encoding
 description: A data compression algorithm for machine translation
```
Comment (bittlingmayer) [inline approaches/byte-pair-encoding.md:6]: Maybe it should be "for tokenization" or "for tokenization for machine translation"?  What do you think?

File: approaches/byte-pair-encoding.md [hunk 3]
```diff
@@ -5,3 +5,25 @@ layout: coming_soon
...
 ---
+
+Byte-pair encoding (BPE) is a data compression algorithm where the most common consecutive pair of bytes is replaced by a byte that is not seen in the data.
```
Comment (cefoo) [inline approaches/byte-pair-encoding.md:9]: Thanks for the article!

To be consistent with other articles, Byte-pair encoding should be formatted in bold.

> **Bite-pair encoding** (BPE) is...

File: approaches/byte-pair-encoding.md [hunk 4]
```diff
@@ -5,3 +5,25 @@ layout: coming_soon
...
+It was first introduced in the article titled "A new Algorithm for Data Compression" by Philip Gage in the February 1994 edition of C Users Journal.
```
Comment (cefoo) [inline approaches/byte-pair-encoding.md:10]: According to the [style guide](https://machinetranslate.org/style#apostrophes-and-quotation-marks), curly quotation marks should be used.

File: approaches/byte-pair-encoding.md [hunk 5]
```diff
@@ -5,3 +5,25 @@ layout: coming_soon
...
+
+A very well-known example of BPE data compression is, suppose we have a string "aaabdaaabac" which is compressed such that, the most frequent byte pair "aa" is replaced by Z. We now have "ZabdZabac". 
+The new most frequent byte pair is "ab" so we replace it by Y such that, we now have "ZYdZYac". 
+Applying recursive byte pair encoding to encode ZY as X, we now have "XdXac". It cannot be compressed further and to decompress the data, replacement in reverse order is performed. 
+
+The use of BPE in NLP ensures that the most common words in the vocabulary are presented as single token whereas the rarest words are broken down into subword tokens instead of introducing it as `<UNK>` tokens.
```
Comment (cefoo) [inline approaches/byte-pair-encoding.md:16]: Although it is understood, please [expand the acronym that hasn't been explained in the article](https://machinetranslate.org/style#acronyms-and-abbreviations). In this case, NLP.

File: approaches/byte-pair-encoding.md (hunk 4 above)
Comment (bittlingmayer) [reply approaches/byte-pair-encoding.md:10]: Maybe for this we should use Italics?

File: approaches/byte-pair-encoding.md [hunk 6]
```diff
@@ -5,3 +5,25 @@ layout: coming_soon
 title: Byte-pair encoding
 description: A data compression algorithm for machine translation
 ---
+
+Byte-pair encoding (BPE) is a data compression algorithm where the most common consecutive pair of bytes is replaced by a byte that is not seen in the data.
+It was first introduced in the article titled "A new Algorithm for Data Compression" by Philip Gage in the February 1994 edition of C Users Journal.
+
+A very well-known example of BPE data compression is, suppose we have a string "aaabdaaabac" which is compressed such that, the most frequent byte pair "aa" is replaced by Z. We now have "ZabdZabac".
```
Comment (bittlingmayer) [inline approaches/byte-pair-encoding.md:12]: We should use backticks for quoting this type of input and output.

File: approaches/byte-pair-encoding.md [hunk 7]
```diff
@@ -5,3 +5,25 @@ layout: coming_soon
...
+
+A very well-known example of BPE data compression is, suppose we have a string "aaabdaaabac" which is compressed such that, the most frequent byte pair "aa" is replaced by Z. We now have "ZabdZabac". 
+The new most frequent byte pair is "ab" so we replace it by Y such that, we now have "ZYdZYac". 
+Applying recursive byte pair encoding to encode ZY as X, we now have "XdXac". It cannot be compressed further and to decompress the data, replacement in reverse order is performed. 
+
+The use of BPE in NLP ensures that the most common words in the vocabulary are presented as single token whereas the rarest words are broken down into subword tokens instead of introducing it as `<UNK>` tokens.
+BPE works well for both character-level and word-level representations and it can manage large corpora.
+It follows greedy approach to find the best possible solution where it goes through every possible option to merge at each iteration by looking at its frequency.
+
+Here are the steps that demonstrates how the algorithm works:
```
Comment (bittlingmayer) [inline approaches/byte-pair-encoding.md:20]: *demonstrate*s, but anyway let's just use *show* or *How the algorithm works:*

**Comment by zouharvi** [comment]
My micro comments:

- Is there a reason for double backticks instead of just one?
- Maybe there could be a part explaining what the output looks like? Just checked and BPE of `praised` is `'__sow', 'pr', 'ai', 'se', 'd', '__eow'` (mind the `'__sow'` and `'__eow'`). I remember being confused at first why they are there.
- Shouldn't there also be a "subwords" article to which BPE and WordPiece link?
- Currently the story is about *compression* (which is in line with the original paper) but the justification for usage in MT is slightly different (exploding vocabulary, typos, etc).

**Comment by bittlingmayer** [comment]
I agree with Vilém on all points.

> Currently the story is about compression (which is in line with the original paper) but the justification for usage in MT is slightly different (exploding vocabulary, typos, etc). Changing the story to "lets begin with one-hot encoded unigrams and see where it fails and why subwords fix it" could be more suitable.

Agree, but I think it's good to introduce it as a compression algorithm, and then - by the 2nd or 3rd sentence - make it clear that it was repurposed for machine translation.

So I would leave the current first or two sentence of the article as-is.

**Comment by bittlingmayer** [comment]
According to the style guide, I think we should:
- use "byte-pair encoding" not "BPE" throughout the article
- use "tokenisation" not "tokenization"

File: approaches/byte-pair-encoding.md [hunk 8]
```diff
@@ -3,5 +3,32 @@ grand_parent: Approaches
...
 layout: coming_soon
 title: Byte-pair encoding
-description: A data compression algorithm for machine translation
+description: A data compression algorithm for tokenisation for machine translation
 ---
+
+Byte-pair encoding (BPE) is a data compression algorithm where the most common consecutive pair of bytes is replaced by a byte that is not seen in the data.
+It was first introduced in the article titled "_A new Algorithm for Data Compression_" by Philip Gage in the February 1994 edition of C Users Journal.
+
+A very well-known example of byte-pair encoding data compression is, suppose we have a string `aaabdaaabac` which is compressed such that, the most frequent byte pair `aa` is replaced by Z. We now have `ZabdZabac`.
```
Comment (cefoo) [inline approaches/byte-pair-encoding.md:12]: To keep content minimal, we can turn this sentence into a step-by-step process. For instance:

```
## Examples

String: `aaabdaaabac`

1- The most frequent byte pair is `aa`, which can be replaced by `Z`. The new string is now `ZabdZabac`.
```

File: approaches/byte-pair-encoding.md [hunk 9]
```diff
@@ -3,5 +3,32 @@ grand_parent: Approaches
...
+Applying recursive byte pair encoding to encode `ZY` as X, we now have `XdXac`. It cannot be compressed further and to decompress the data, replacement in reverse order is performed. 
+
+The use of byte-pair encoding in NLP ensures that the most common words in the vocabulary are presented as single token whereas the rarest words are broken down into subword tokens instead of introducing it as `<UNK>` tokens.
+Byte-pair encoding works well for both character-level and word-level representations and it can manage large corpora.
+It follows greedy approach to find the best possible solution where it goes through every possible option to merge at each iteration by looking at its frequency.
+
+Implementation of byte-pair encoding is slightly modified to fit for subword tokenisation.
+Here, frequently occuring subwords are merged instead of being replaced by another byte as it was in compression.
+For example, let's consider the following unigram `praise`. It is split up into more frequent words and are represented as `['__sow', 'pr', 'ai', 'se', 'd', '__eow']`, where `'__sow'` and `'__eow'` are the markers for start and end of the tokens.
```
Comment (cefoo) [inline approaches/byte-pair-encoding.md:22]: We can copy the structure of the previous example here and break down the example into steps.

File: approaches/byte-pair-encoding.md [hunk 10]
```diff
@@ -3,5 +3,32 @@ grand_parent: Approaches
...
+
+Below are the steps that shows how the algorithm works:
+
+1. Initialise the vocabulary with the frequency of occurence of words in the corpus.
+2. Represent each word in the corpus as a concatenation with special token `</w>`.
+3. Split each word into character and count their occurence.
+4. Look for the most frequent pairing, merge them and perform the same iteration again.
+5. Repeat 4 until all the iterations are performed or the token limit size is reached.
+
+Byte-pair encoding is a widely used subword-tokenisation algorithms.
```
Comment (cefoo) [inline approaches/byte-pair-encoding.md:33]: Maybe these last two sentences can be omitted to keep the article shorter.

File: approaches/byte-pair-encoding.md [hunk 11]
```diff
@@ -3,5 +3,32 @@ grand_parent: Approaches
 parent: Neural machine translation
 layout: coming_soon
```
Comment (cefoo) [inline approaches/byte-pair-encoding.md:4]: The `coming_soon` label is now outdated. 🥳 You can delete that line. 😀

**Review by cefoo — COMMENTED** [review]
Hi Dali! 🌼

Thanks a lot for the updates!

I've left some suggestions to your PR. Please let me know what you think.

All in all, the article is great and super clear!

**Pull request merged**